java/src/main/java/com/google/appengine/tools/pipeline/impl/backend/AppEngineServicesServiceImpl.java

```diff
@@ -0,0 +1,103 @@
+package com.google.appengine.tools.pipeline.impl.backend;
+
+import com.google.appengine.v1.*;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import lombok.NonNull;
+import lombok.SneakyThrows;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+
+public class AppEngineServicesServiceImpl implements AppEngineServicesService {
+
+  private final AppEngineEnvironment appEngineEnvironment;
+
+  private final Provider<ServicesClient>  servicesClientProvider;
+  private final Provider<VersionsClient>  versionsClientProvider;
+
+  @Inject
+  AppEngineServicesServiceImpl(AppEngineEnvironment appEngineEnvironment,
+                               Provider<ServicesClient> servicesClientProvider,
+                               Provider<VersionsClient> versionsClientProvider) {
+    this.appEngineEnvironment = appEngineEnvironment;
+    this.servicesClientProvider = servicesClientProvider;
+    this.versionsClientProvider = versionsClientProvider;
+  }
+
+  @SneakyThrows
+  public static AppEngineServicesServiceImpl defaults() {
+    return new AppEngineServicesServiceImpl(new AppEngineStandardGen2(), AppEngineServicesServiceImpl::getServicesClientProvider, AppEngineServicesServiceImpl::getVersionsClientProvider);
+  }
+
+  static ServicesClient getServicesClientProvider() {
```

Comment:
This and `getVersionsClientProvider` seems to be only used in the test and in the default constructor. Ok for the default constructor if we are skipping any DI there, but I'd moved as private and leave test to be constructed with the other constructor, which is the one is being used.